app/src/main/java/org/vss/impl/postgres/PostgresBackendImpl.java

```diff
@@ -127,6 +126,48 @@ private VssDbRecord buildVssRecord(String storeId, KeyValue kv) {
 
   @Override
   public ListKeyVersionsResponse listKeyVersions(ListKeyVersionsRequest request) {
-    throw new UnsupportedOperationException("Operation not implemented");
+    String storeId = request.getStoreId();
+    String keyPrefix = request.getKeyPrefix();
+    String pageToken = request.getPageToken();
+    int pageSize = request.getPageSize();
+
+    long globalVersion = 0L;
+
+    // Only fetch global_version for first page.
+    // Fetch global_version before fetching any key_versions to ensure that,
+    // all current key_versions were stored at global_version or later.
+    if (!request.hasPageToken()) {
+      globalVersion = get(GetObjectRequest.newBuilder()
+          .setStoreId(storeId)
+          .setKey(GLOBAL_VERSION_KEY)
+          .build()).getValue().getVersion();
+    }
+
+    List<VssDbRecord> vssDbRecords = context.select(VSS_DB.KEY, VSS_DB.VERSION).from(VSS_DB)
+        .where(VSS_DB.STORE_ID.eq(storeId)
+            .and(VSS_DB.KEY.startsWith(keyPrefix)))
+        .orderBy(VSS_DB.KEY)
+        .seek(pageToken)
+        .limit(Math.min(pageSize, 50))
+        .getResult().map(record -> record.into(VssDbRecord.class));
+
+    List<KeyValue> keyVersions = vssDbRecords.stream()
+        .filter(kv -> !GLOBAL_VERSION_KEY.equals(kv.getKey()))
```

Comment:
Yes, it can mean number of entries in response can be one less than pageSize. 
But that shouldn't be a concern and client should not assume response to contain specific number of entries. 
For e.g. max number of results in paginated response can change at anytime with no notice to client.

We already caution against this in api doc in proto: 
"Caution: Clients must not assume a specific number of key_versions to be present in a page for paginated response."
Only way to know whether nextPage exists or not is by presence of nextPageToken.

Reply:
That doesn't answer my second question. :)

> Should we filter at the SQL level instead?

Is there a reason not to?

Reply:
There is no big reason in case of SQL, i can do that. (apart from client/api-expectation and precedent) 
It is just something to keep in mind that this operation might not be supported by all KV-database i.e. (list along with key not equals).

Reply:
Thanks ok. Feel fee to leave it as is.